cdap-client/src/main/java/co/cask/cdap/client/ProgramClient.java

```diff
@@ -401,6 +407,36 @@ public String call() throws Exception {
   }
 
   /**
+   * Waits for a program to have a certain status.
+   *
+   * @param program the program
+   * @param status the desired status
+   * @param timeout how long to wait in milliseconds until timing out
+   * @throws IOException if a network error occurred
+   * @throws ProgramNotFoundException if the program with the specified name could not be found
+   * @throws UnauthenticatedException if the request is not authorized successfully in the gateway server
+   * @throws TimeoutException if the program did not achieve the desired program status before the timeout
+   * @throws InterruptedException if interrupted while waiting for the desired program status
+   */
+  public void waitForStatus(final ProgramId program, String status, long timeout, TimeUnit timeoutUnit)
+    throws UnauthenticatedException, IOException, ProgramNotFoundException,
+    TimeoutException, InterruptedException {
+
+    try {
+      Tasks.waitFor(status, new Callable<String>() {
+        @Override
+        public String call() throws Exception {
+          return getStatus(program);
+        }
+      }, timeout, timeoutUnit, 1, TimeUnit.SECONDS);
+    } catch (ExecutionException e) {
+      Throwables.propagateIfPossible(e.getCause(), UnauthenticatedException.class);
+      Throwables.propagateIfPossible(e.getCause(), ProgramNotFoundException.class);
+      Throwables.propagateIfPossible(e.getCause(), IOException.class);
```

Comment:
I understand that the existing waitForStatus logic also does this, but... this seems wrong.
What if the `getCause` returns some other exception besides the 3 listed here? Then it will just suppress the exception.